### DISCUSSION TITLE
Execute logica over pandas dataframes

### DISCUSSION THREAD

**Discussion by wdduncan** [EvgSkv/logica#305] [Ideas]
Is is possible to execute Logica on pandas dataframes? 
If not, this would be a nice enhancement.

**Comment by EvgSkv** [comment]
We are thinking of this actually. We expect to have this feature by June 2024, hopefully much sooner.

**Reply by wdduncan** [reply]
Great to hear!

**Reply by EvgSkv** [reply]
@wdduncan You can run Logica over Pandas dataframes now! 😁 🎉 
[EXAMPLE](https://colab.research.google.com/drive/11eXl46O0ZWc_O4pX4BaDAkjfRFf8M0n_?usp=sharing)

<img width="826" alt="PandasLogica" src="https://github.com/EvgSkv/logica/assets/71790359/bbf1716d-cc2b-4a5b-bb4a-223d1e048f63">
Please try it out when you have a chance and let me know if you run into any problems!

**Reply by wdduncan** [reply]
Most EXCELLENT!  
Thanks @EvgSkv :)

**Reply by wdduncan** [reply]
@EvgSkv I am trying to use `logica_lib.RunPredicateToPandas` to query a pandas dataframe. Following your example, here is what I have so far:
```python
xs = numpy.linspace(1.0, 6.0, 100000000)
ys = numpy.sin(xs)
sin_wave = pandas.DataFrame({'x': xs, 'y': ys})

logica_lib.RunPredicateToPandas("""
@Engine("sqlite");
Pi() Min= x :- sin_wave(x:, y:), y < 0;
""", 'Pi')
```
However, it can't the `sin_wave` dataframe. It returns error:
```python
FileNotFoundError: [Errno 2] No such file or directory: '\n@Engine("sqlite");\nPi() Min= x :- sin_wave(x:, y:), y < 0;\n'
```

**Reply by EvgSkv** [reply]
There was actually a bug :-)

Please try now:

```
import numpy
import pandas
import duckdb

xs = numpy.linspace(1.0, 6.0, 100000000)
ys = numpy.sin(xs)
sin_wave = pandas.DataFrame({'x': xs, 'y': ys})
logica_lib.sin_wave = sin_wave

logica_lib.RunPredicateFromString("""
@Engine("duckdb");
Pi() Min= x :- sin_wave(x:, y:), y < 0;
""", 'Pi', connection=duckdb.connect())
```

Note the line: `logica_lib.sin_wave = sin_wave`, this is required for duckdb function call to see the `sin_wave` in the Python tree.

**Reply by wdduncan** [reply]
Thanks!  
Can you make a new release of `logica`. Your code returns error:
```python
Exception: Logica only supports BigQuery, PostgreSQL and SQLite for now.
```

**Reply by EvgSkv** [reply]
Sorry for delay, this will be done in the nearest days.

**Reply by EvgSkv** [reply]
I just realized that I didn't give a link to the colab demo:
https://colab.research.google.com/drive/1yfDVhssG1B91u_u1m_cQldEIFUKOk0Mw?usp=sharing

Also I made a release, so should work from pip now:
https://colab.research.google.com/drive/1poxzUt-Hu1kLSrKbaodALGGAcZOCL8bs?usp=sharing

Let me know if you hit any more issues!

**Reply by wdduncan** [reply]
Thanks @EvgSkv !
